### DISCUSSION TITLE
What is a minimalistic set of top-level dependencies which is not cross-platform?

### DISCUSSION THREAD

**Discussion by stfdxv** [pdm-project/pdm#1933] [Q&A] [answered]
One of PDM's strong points is the ability to make cross-platform lock-files. If I understand it correctly, the same cross-platformness can't be achieved by one requirements.txt made via `pip freeze` or `pip-compile`.
So what is some simple example of a dependency set, which has incompatible requirements.txt on Linux and Windows (but can be locked by PDM)?

**Comment by frostming** [comment, marked as answer]
Here is a single package that has conditional dependencies: `click`

**Reply by stfdxv** [reply]
I see. In the case of `click` its dependency - `colorama` - will be installed only on Windows. But it's still possible to use requirements.txt made on Windows for other platforms.
Are there any cases where requirements.txt files on different platforms are absolutely incompatible? For example, different versions of the same dependency are used on Windows and Linux.

**Reply by Drachenfels** [reply]
Are there any cases? For sure, you can make one, but platform-dependent requirements are quite rare as python tends to be quite agnostic, also there are ways to avoid it (like repackaging 3rd party packages, forcing installs anyway, manipulating pythonpath, etc).

It is worth perhaps mentioning that pdm export is not exporting lockfile, so platform qualifiers are missing.

If you want to try to break it yourself (for science) try to force specific version of dependency (say greenlet 0.4.17) and try package that does not allow given version (say `pdm add "sqlalchemy==2.0.13"`.

Other notable examples I can dig out from pdm.lock that show some other interesting qualifiers:

```
[[package]]
name = "sqlalchemy"
version = "2.0.13"
requires_python = ">=3.7"
summary = "Database Abstraction Library"
dependencies = [
    "greenlet!=0.4.17; platform_machine == \"aarch64\" or (platform_machine == \"ppc64le\" or (platform_machine == \"x86_64\" or (platform_machine == \"amd64\" or (platform_machine == \"AMD64\" or (platform_machine == \"win32\" or platform_machine == \"WIN32\")))))",
    "typing-extensions>=4.2.0",
]
```

```
[[package]]
name = "pytest"
version = "7.3.1"
requires_python = ">=3.7"
summary = "pytest: simple powerful testing with Python"
dependencies = [
    "colorama; sys_platform == \"win32\"",
    "exceptiongroup>=1.0.0rc8; python_version < \"3.11\"",
    "iniconfig",
    "packaging",
    "pluggy<2.0,>=0.12",
    "tomli>=1.0.0; python_version < \"3.11\"",
]
```

```
[[package]]
name = "virtualenv"
version = "20.23.0"
requires_python = ">=3.7"
summary = "Virtual Python Environment builder"
dependencies = [
    "distlib<1,>=0.3.6",
    "filelock<4,>=3.11",
    "platformdirs<4,>=3.2",
]
```

**Reply by stfdxv** [reply]
@Drachenfels thanks for your comment. I guess you are right: a non-cross-platform dependency set is not that common as I initially imagined.
On the second thought, the `click` package proposed by Frost Ming does count as the requested example. It occurs naturally and leads to a missing dependency if you `pip freeze > req.txt` on Linux and then `pip install -r req.txt` on Windows.
